ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java

```diff
@@ -0,0 +1,226 @@
+        mockWithAction(() -> downloadState.addOutstandingTask(mock(EthTask.class)));
+
+    downloadState.setWaitingForNewPeer(true);
+    downloadState.whileAdditionalRequestCanBeSent(mustNotBeCalled());
+
+    downloadState.setWaitingForNewPeer(false);
+    downloadState.whileAdditionalRequestCanBeSent(sendRequest);
+    verify(sendRequest, times(MAX_OUTSTANDING_REQUESTS)).run();
+  }
+
+  @Test
+  public void shouldStopSendingAdditionalRequestsWhenPendingRequestsIsEmpty() {
+    pendingRequests.enqueue(createAccountDataRequest(Hash.EMPTY_TRIE_HASH));
+    pendingRequests.enqueue(createAccountDataRequest(Hash.EMPTY_TRIE_HASH));
+
+    final Runnable sendRequest = mockWithAction(pendingRequests::dequeue);
+    downloadState.whileAdditionalRequestCanBeSent(sendRequest);
+
+    verify(sendRequest, times(2)).run();
+  }
+
+  @Test
+  public void shouldStopSendingAdditionalRequestsWhenMaximumOutstandingRequestCountReached() {
+    pendingRequests.enqueue(createAccountDataRequest(Hash.EMPTY_TRIE_HASH));
+    final Runnable sendRequest =
+        mockWithAction(() -> downloadState.addOutstandingTask(mock(EthTask.class)));
+
+    downloadState.whileAdditionalRequestCanBeSent(sendRequest);
+    verify(sendRequest, times(MAX_OUTSTANDING_REQUESTS)).run();
+  }
+
+  @Test
+  public void shouldStopSendingAdditionalRequestsWhenFutureIsCancelled() {
+    pendingRequests.enqueue(createAccountDataRequest(Hash.EMPTY_TRIE_HASH));
+    final Runnable sendRequest = mockWithAction(() -> future.cancel(true));
+
+    downloadState.whileAdditionalRequestCanBeSent(sendRequest);
+    verify(sendRequest, times(1)).run();
+  }
+
+  @Test
+  public void shouldStopSendingAdditionalRequestsWhenDownloadIsMarkedAsStalled() {
+    pendingRequests.enqueue(createAccountDataRequest(Hash.EMPTY_TRIE_HASH));
+    final Runnable sendRequest = mockWithAction(() -> downloadState.markAsStalled(1));
+
+    downloadState.whileAdditionalRequestCanBeSent(sendRequest);
+    verify(sendRequest, times(1)).run();
+  }
+
+  @Test
+  public void shouldNotAllowMultipleCallsToSendAdditionalRequestsAtOnce() {
+    pendingRequests.enqueue(createAccountDataRequest(Hash.EMPTY_TRIE_HASH));
+    final Runnable sendRequest =
+        mockWithAction(
+            () -> {
+              downloadState.whileAdditionalRequestCanBeSent(mustNotBeCalled());
+              downloadState.addOutstandingTask(mock(EthTask.class));
+            });
+
+    downloadState.whileAdditionalRequestCanBeSent(sendRequest);
+    verify(sendRequest, times(MAX_OUTSTANDING_REQUESTS)).run();
+  }
+
+  @Test
+  public void shouldNotEnqueueRequestsAfterDownloadIsStalled() {
+    downloadState.checkCompletion(worldStateStorage, header);
+
+    downloadState.enqueueRequests(Stream.of(createAccountDataRequest(Hash.EMPTY_TRIE_HASH)));
+    downloadState.enqueueRequest(createAccountDataRequest(Hash.EMPTY_TRIE_HASH));
+
+    assertThat(pendingRequests.isEmpty()).isTrue();
+  }
+
+  private Runnable mustNotBeCalled() {
```

Comment:
Putting on my paranoid tin foil hat, do we have a test where mustNotBeCalled is called so we can assert the failure escapes the lambdas?  Or maybe I'm being too paranoid.

Reply:
I was going to say you were paranoid but Meredith had the same question.  I know it works because I always check my tests fail but worth adding an explicit test.